// @testmo/testmo-cli/dist/parsers/junit.js
"use strict";
/*
 * Copyright (c) Testmo GmbH (Berlin, Germany)
 * All rights reserved.
 * contact@testmo.com - www.testmo.com
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JUnitParser = void 0;
var parser_1 = require("../core/parser");
var fast_xml_parser_1 = __importDefault(require("fast-xml-parser"));
var validator_1 = __importDefault(require("validator"));
var path_1 = __importDefault(require("path"));
var lodash_1 = __importDefault(require("lodash"));
var he_1 = __importDefault(require("he"));
var JUnitParser = /** @class */ (function (_super) {
    __extends(JUnitParser, _super);
    function JUnitParser() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = 'JUnit XML';
        return _this;
    }
    JUnitParser.prototype.detect = function (content, filename) {
        if (path_1.default.extname(filename).toLowerCase() != '.xml') {
            return false;
        }
        // Some tools generate XML files without leading <?xml>, so we
        // allow different checks
        return /<\?xml.+<testsuite/s.test(content) ||
            /<testsuites.+<testsuite/s.test(content);
    };
    JUnitParser.prototype.hasNonEmptyAttribute = function (element, attribute) {
        return (attribute in element) &&
            Array.isArray(element[attribute]) &&
            (element[attribute].length > 0) &&
            !validator_1.default.isEmpty(element[attribute][0]);
    };
    JUnitParser.prototype.getTimeAttribute = function (element, attribute) {
        var value = element[attribute][0].replace(/,/, '.');
        if (!validator_1.default.isNumeric(value)) {
            return;
        }
        var time = Number.parseFloat(value);
        if (Number.isNaN(time) || time < 0) {
            return;
        }
        // Convert to microseconds
        return Math.round(time * 1000000);
    };
    JUnitParser.prototype.convertTextElementToObject = function (element) {
        if (typeof element === 'string' && !validator_1.default.isEmpty(element)) {
            return {
                '#text': [element]
            };
        }
        return element;
    };
    JUnitParser.prototype.getIntAttribute = function (element, attribute) {
        var value = element[attribute][0];
        if (!validator_1.default.isInt(value)) {
            return;
        }
        var intValue = Number.parseInt(value, 10);
        if (Number.isNaN(intValue)) {
            return;
        }
        return intValue;
    };
    JUnitParser.prototype.getTopElement = function (xml, evaluate) {
        var recursiveGetTopElement = function (element) {
            if (!(element instanceof Object)) {
                return;
            }
            if (evaluate(element)) {
                return element;
            }
            // Does this element have just a single testsuite or testsuites
            // child and not testcase elements?
            var testsuitesChildren = 0;
            var testsuiteChildren = 0;
            var testcaseChildren = 0;
            if ('testsuites' in element && Array.isArray(element.testsuites)) {
                testsuitesChildren = element.testsuites.length;
            }
            if ('testsuite' in element && Array.isArray(element.testsuite)) {
                testsuiteChildren = element.testsuite.length;
            }
            if ('testcase' in element && Array.isArray(element.testcase)) {
                testcaseChildren = element.testcase.length;
            }
            // We got a node with test cases, so any testsuite or testsuites childs
            // cannot be the top element anymore
            if (testcaseChildren > 0) {
                return;
            }
            if (testsuitesChildren == 1 && testsuiteChildren == 0) {
                return recursiveGetTopElement(element.testsuites[0]);
            }
            if (testsuitesChildren == 0 && testsuiteChildren == 1) {
                return recursiveGetTopElement(element.testsuite[0]);
            }
        };
        return recursiveGetTopElement(xml);
    };
    JUnitParser.prototype.getFieldsForElement = function (element) {
        var fields = [];
        var outputFields = {
            'system-out': 'Output',
            'system-err': 'Errors'
        };
        for (var outputField in outputFields) {
            if (outputField in element && Array.isArray(element[outputField])) {
                var output = this.convertTextElementToObject(element[outputField][0]);
                if (output instanceof Object && '#text' in output) {
                    fields.push({
                        source: outputField == 'system-err' ? 'stderr' : 'stdout',
                        type: 'console',
                        name: outputFields[outputField],
                        value: output['#text'][0],
                        isHighlight: false,
                        meta: {}
                    });
                }
            }
        }
        if ('properties' in element && Array.isArray(element.properties) &&
            element.properties.length >= 1) {
            var properties = element.properties[0];
            if (properties instanceof Object &&
                'property' in properties && Array.isArray(properties.property)) {
                var propertyList = properties.property;
                for (var _i = 0, propertyList_1 = propertyList; _i < propertyList_1.length; _i++) {
                    var property = propertyList_1[_i];
                    if (property instanceof Object &&
                        this.hasNonEmptyAttribute(property, '@_name') &&
                        this.hasNonEmptyAttribute(property, '@_value')) {
                        fields.push({
                            source: 'property',
                            type: 'string',
                            name: property['@_name'][0],
                            value: property['@_value'][0],
                            isHighlight: false,
                            meta: {}
                        });
                    }
                }
            }
        }
        return fields;
    };
    JUnitParser.prototype.collectTests = function (xml) {
        var _this = this;
        var getTestResultStatusField = function (element, status) {
            // We only add a new field if we get some useful information about
            // the status. E.g. for a simple <skipped/> element without content
            // and without attributes, we don't add a field.
            element = _this.convertTextElementToObject(element);
            if (!(element instanceof Object)) {
                // Empty tag
                return;
            }
            var meta = {};
            var text = '#text' in element ? element['#text'][0] : undefined;
            var value = text;
            // Use attribute for value priority: message > type
            for (var _i = 0, _a = ['message', 'type']; _i < _a.length; _i++) {
                var attribute = _a[_i];
                if (_this.hasNonEmptyAttribute(element, '@_' + attribute)) {
                    var attributeValue = element['@_' + attribute][0];
                    if (value === undefined) {
                        value = attributeValue;
                    }
                    else {
                        meta[attribute] = attributeValue;
                    }
                }
            }
            // We couldn't find any useful information
            if (value === undefined) {
                return;
            }
            return {
                source: 'status',
                type: 'console',
                name: lodash_1.default.startCase(status.toLowerCase()),
                value: value,
                isHighlight: true,
                meta: meta
            };
        };
        var addTestAttributes = function (test, element) {
            if (_this.hasNonEmptyAttribute(element, '@_classname')) {
                test.folder = element['@_classname'][0];
            }
            else if (_this.hasNonEmptyAttribute(element, '@_class')) {
                test.folder = element['@_class'][0];
            }
            if (_this.hasNonEmptyAttribute(element, '@_time')) {
                var time = _this.getTimeAttribute(element, '@_time');
                if (time !== undefined) {
                    test.elapsed = time;
                }
            }
            if (test.elapsed === undefined &&
                _this.hasNonEmptyAttribute(element, '@_duration')) {
                var duration = _this.getTimeAttribute(element, '@_duration');
                if (duration !== undefined) {
                    test.elapsed = duration;
                }
            }
            if (_this.hasNonEmptyAttribute(element, '@_file')) {
                test.file = element['@_file'][0];
            }
            if (_this.hasNonEmptyAttribute(element, '@_line')) {
                var line = _this.getIntAttribute(element, '@_line');
                if (line !== undefined && line >= 0) {
                    test.line = line;
                }
            }
            if (_this.hasNonEmptyAttribute(element, '@_assertions')) {
                var assertions = _this.getIntAttribute(element, '@_assertions');
                if (assertions !== undefined && assertions >= 0) {
                    test.assertions = assertions;
                }
            }
        };
        var getTest = function (element) {
            var _a;
            if (!(element instanceof Object)) {
                // Empty test case
                return;
            }
            if (!_this.hasNonEmptyAttribute(element, '@_name')) {
                return;
            }
            var test = {
                key: element['@_name'][0],
                name: element['@_name'][0],
                status: 'passed',
                fields: []
            };
            addTestAttributes(test, element);
            // Test status priority: error > failure > skipped
            for (var _i = 0, _b = ['skipped', 'failure', 'error']; _i < _b.length; _i++) {
                var status_1 = _b[_i];
                if (status_1 in element) {
                    for (var _c = 0, _d = element[status_1]; _c < _d.length; _c++) {
                        var statusElement = _d[_c];
                        var field = getTestResultStatusField(statusElement, status_1);
                        if (field !== undefined) {
                            test.fields.push(field);
                        }
                    }
                    test.status = status_1;
                }
            }
            (_a = test.fields).push.apply(_a, _this.getFieldsForElement(element));
            // Improve test key if possible
            if ('folder' in test) {
                test.key = "".concat(test.folder, ".").concat(test.name);
            }
            return test;
        };
        var getTestsForSuite = function (parentElement) {
            if (!(parentElement instanceof Object)) {
                return [];
            }
            var tests = [];
            if ('testsuite' in parentElement) {
                for (var _i = 0, _a = parentElement.testsuite; _i < _a.length; _i++) {
                    var testsuite = _a[_i];
                    tests.push.apply(tests, getTestsForSuite(testsuite));
                }
            }
            if ('testsuites' in parentElement) {
                for (var _b = 0, _c = parentElement.testsuites; _b < _c.length; _b++) {
                    var testsuites = _c[_b];
                    tests.push.apply(tests, getTestsForSuite(testsuites));
                }
            }
            if ('testcase' in parentElement) {
                for (var _d = 0, _e = parentElement.testcase; _d < _e.length; _d++) {
                    var testcase = _e[_d];
                    var test_1 = getTest(testcase);
                    if (test_1 !== undefined) {
                        tests.push(test_1);
                    }
                }
            }
            return tests;
        };
        return getTestsForSuite(xml);
    };
    JUnitParser.prototype.collectElapsed = function (xml) {
        if (!(xml instanceof Object)) {
            return;
        }
        var topElement = this.getTopElement(xml, function (element) {
            return ('@_time' in element) || ('@_duration' in element);
        });
        if (topElement === undefined) {
            return;
        }
        if (this.hasNonEmptyAttribute(topElement, '@_time')) {
            var time = this.getTimeAttribute(topElement, '@_time');
            if (time !== undefined) {
                return time;
            }
        }
        if (this.hasNonEmptyAttribute(topElement, '@_duration')) {
            var duration = this.getTimeAttribute(topElement, '@_duration');
            if (duration !== undefined) {
                return duration;
            }
        }
    };
    JUnitParser.prototype.collectFields = function (xml) {
        if (!(xml instanceof Object)) {
            return [];
        }
        var topElement = this.getTopElement(xml, function (element) {
            return ('properties' in element) ||
                ('system-out' in element) ||
                ('system-err' in element);
        });
        if (topElement === undefined) {
            return [];
        }
        return this.getFieldsForElement(topElement);
    };
    JUnitParser.prototype.convertToResults = function (xml) {
        var tests = this.collectTests(xml);
        var fields = this.collectFields(xml);
        var results = {
            tests: tests,
            fields: fields
        };
        var elapsed = this.collectElapsed(xml);
        if (elapsed !== undefined) {
            results.elapsed = elapsed;
        }
        return results;
    };
    JUnitParser.prototype.parse = function (content) {
        var xml = fast_xml_parser_1.default.parse(content, {
            arrayMode: 'strict',
            ignoreAttributes: false,
            parseAttributeValue: false,
            parseNodeValue: false,
            attrValueProcessor: function (val, attrName) { return he_1.default.decode(val, { isAttributeValue: true }); },
            tagValueProcessor: function (val, tagName) { return he_1.default.decode(val); },
        }, true);
        return this.convertToResults(xml);
    };
    return JUnitParser;
}(parser_1.Parser));
exports.JUnitParser = JUnitParser;
