// @testmo/testmo-api/dist/model/CreateAutomationRunThread.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _ApiClient = _interopRequireDefault(require("../ApiClient"));

var _CreateAutomationArtifact = _interopRequireDefault(require("./CreateAutomationArtifact"));

var _CreateAutomationField = _interopRequireDefault(require("./CreateAutomationField"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

/**
 * The CreateAutomationRunThread model module.
 * @module model/CreateAutomationRunThread
 * @version 1.0.0
 */
var CreateAutomationRunThread = /*#__PURE__*/function () {
  /**
   * Constructs a new <code>CreateAutomationRunThread</code>.
   * @alias module:model/CreateAutomationRunThread
   */
  function CreateAutomationRunThread() {
    _classCallCheck(this, CreateAutomationRunThread);

    CreateAutomationRunThread.initialize(this);
  }
  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */


  _createClass(CreateAutomationRunThread, null, [{
    key: "initialize",
    value: function initialize(obj) {}
    /**
     * Constructs a <code>CreateAutomationRunThread</code> from a plain JavaScript object, optionally creating a new instance.
     * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
     * @param {Object} data The plain JavaScript object bearing properties of interest.
     * @param {module:model/CreateAutomationRunThread} obj Optional instance to populate.
     * @return {module:model/CreateAutomationRunThread} The populated <code>CreateAutomationRunThread</code> instance.
     */

  }, {
    key: "constructFromObject",
    value: function constructFromObject(data, obj) {
      if (data) {
        obj = obj || new CreateAutomationRunThread();

        if (data.hasOwnProperty('elapsed_observed')) {
          obj['elapsed_observed'] = _ApiClient["default"].convertToType(data['elapsed_observed'], 'Number');
        }

        if (data.hasOwnProperty('elapsed_computed')) {
          obj['elapsed_computed'] = _ApiClient["default"].convertToType(data['elapsed_computed'], 'Number');
        }

        if (data.hasOwnProperty('artifacts')) {
          obj['artifacts'] = _ApiClient["default"].convertToType(data['artifacts'], [_CreateAutomationArtifact["default"]]);
        }

        if (data.hasOwnProperty('fields')) {
          obj['fields'] = _ApiClient["default"].convertToType(data['fields'], [_CreateAutomationField["default"]]);
        }
      }

      return obj;
    }
  }]);

  return CreateAutomationRunThread;
}();
/**
 * Observed overall elapsed (execution time) of the thread in microseconds.
 * @member {Number} elapsed_observed
 */


CreateAutomationRunThread.prototype['elapsed_observed'] = undefined;
/**
 * Computed overall elapsed (execution time) of the thread in microseconds.  The _observed_ elapsed is usually derived from measuring the execution time of tests whereas the _computed_ elapsed is calculated from result log files.
 * @member {Number} elapsed_computed
 */

CreateAutomationRunThread.prototype['elapsed_computed'] = undefined;
/**
 * List of externally stored test artifacts to link to the new thread (such as log files, screenshots or test data).
 * @member {Array.<module:model/CreateAutomationArtifact>} artifacts
 */

CreateAutomationRunThread.prototype['artifacts'] = undefined;
/**
 * List of fields to attach to the new thread (such as environment variables, errors or terminal output).
 * @member {Array.<module:model/CreateAutomationField>} fields
 */

CreateAutomationRunThread.prototype['fields'] = undefined;
var _default = CreateAutomationRunThread;
exports["default"] = _default;
