// @testmo/testmo-api/dist/model/CreateAutomationLink.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _ApiClient = _interopRequireDefault(require("../ApiClient"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

/**
 * The CreateAutomationLink model module.
 * @module model/CreateAutomationLink
 * @version 1.0.0
 */
var CreateAutomationLink = /*#__PURE__*/function () {
  /**
   * Constructs a new <code>CreateAutomationLink</code>.
   * @alias module:model/CreateAutomationLink
   * @param name {String} Name or file name of the link.
   * @param url {String} Link to an external resource or website.
   */
  function CreateAutomationLink(name, url) {
    _classCallCheck(this, CreateAutomationLink);

    CreateAutomationLink.initialize(this, name, url);
  }
  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */


  _createClass(CreateAutomationLink, null, [{
    key: "initialize",
    value: function initialize(obj, name, url) {
      obj['name'] = name;
      obj['url'] = url;
    }
    /**
     * Constructs a <code>CreateAutomationLink</code> from a plain JavaScript object, optionally creating a new instance.
     * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
     * @param {Object} data The plain JavaScript object bearing properties of interest.
     * @param {module:model/CreateAutomationLink} obj Optional instance to populate.
     * @return {module:model/CreateAutomationLink} The populated <code>CreateAutomationLink</code> instance.
     */

  }, {
    key: "constructFromObject",
    value: function constructFromObject(data, obj) {
      if (data) {
        obj = obj || new CreateAutomationLink();

        if (data.hasOwnProperty('name')) {
          obj['name'] = _ApiClient["default"].convertToType(data['name'], 'String');
        }

        if (data.hasOwnProperty('note')) {
          obj['note'] = _ApiClient["default"].convertToType(data['note'], 'String');
        }

        if (data.hasOwnProperty('url')) {
          obj['url'] = _ApiClient["default"].convertToType(data['url'], 'String');
        }
      }

      return obj;
    }
  }]);

  return CreateAutomationLink;
}();
/**
 * Name or file name of the link.
 * @member {String} name
 */


CreateAutomationLink.prototype['name'] = undefined;
/**
 * Short note or summary of the link with a maximum of 80 characters.
 * @member {String} note
 */

CreateAutomationLink.prototype['note'] = undefined;
/**
 * Link to an external resource or website.
 * @member {String} url
 */

CreateAutomationLink.prototype['url'] = undefined;
var _default = CreateAutomationLink;
exports["default"] = _default;
