// @testmo/testmo-api/dist/model/CreateAutomationRun.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _ApiClient = _interopRequireDefault(require("../ApiClient"));

var _CreateAutomationArtifact = _interopRequireDefault(require("./CreateAutomationArtifact"));

var _CreateAutomationField = _interopRequireDefault(require("./CreateAutomationField"));

var _CreateAutomationLink = _interopRequireDefault(require("./CreateAutomationLink"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

/**
 * The CreateAutomationRun model module.
 * @module model/CreateAutomationRun
 * @version 1.0.0
 */
var CreateAutomationRun = /*#__PURE__*/function () {
  /**
   * Constructs a new <code>CreateAutomationRun</code>.
   * @alias module:model/CreateAutomationRun
   * @param name {String} Name of the new automation run.
   * @param source {String} Name of the source for the new automation run. If this source does not already exist in the target project, Testmo automatically creates a new one.  It is recommended to keep source names short. Good examples are `backend`, `frontend` or `mobile-iphone`.
   */
  function CreateAutomationRun(name, source) {
    _classCallCheck(this, CreateAutomationRun);

    CreateAutomationRun.initialize(this, name, source);
  }
  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */


  _createClass(CreateAutomationRun, null, [{
    key: "initialize",
    value: function initialize(obj, name, source) {
      obj['name'] = name;
      obj['source'] = source;
    }
    /**
     * Constructs a <code>CreateAutomationRun</code> from a plain JavaScript object, optionally creating a new instance.
     * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
     * @param {Object} data The plain JavaScript object bearing properties of interest.
     * @param {module:model/CreateAutomationRun} obj Optional instance to populate.
     * @return {module:model/CreateAutomationRun} The populated <code>CreateAutomationRun</code> instance.
     */

  }, {
    key: "constructFromObject",
    value: function constructFromObject(data, obj) {
      if (data) {
        obj = obj || new CreateAutomationRun();

        if (data.hasOwnProperty('name')) {
          obj['name'] = _ApiClient["default"].convertToType(data['name'], 'String');
        }

        if (data.hasOwnProperty('source')) {
          obj['source'] = _ApiClient["default"].convertToType(data['source'], 'String');
        }

        if (data.hasOwnProperty('config')) {
          obj['config'] = _ApiClient["default"].convertToType(data['config'], 'String');
        }

        if (data.hasOwnProperty('config_id')) {
          obj['config_id'] = _ApiClient["default"].convertToType(data['config_id'], 'Number');
        }

        if (data.hasOwnProperty('milestone')) {
          obj['milestone'] = _ApiClient["default"].convertToType(data['milestone'], 'String');
        }

        if (data.hasOwnProperty('milestone_id')) {
          obj['milestone_id'] = _ApiClient["default"].convertToType(data['milestone_id'], 'Number');
        }

        if (data.hasOwnProperty('tags')) {
          obj['tags'] = _ApiClient["default"].convertToType(data['tags'], ['String']);
        }

        if (data.hasOwnProperty('artifacts')) {
          obj['artifacts'] = _ApiClient["default"].convertToType(data['artifacts'], [_CreateAutomationArtifact["default"]]);
        }

        if (data.hasOwnProperty('fields')) {
          obj['fields'] = _ApiClient["default"].convertToType(data['fields'], [_CreateAutomationField["default"]]);
        }

        if (data.hasOwnProperty('links')) {
          obj['links'] = _ApiClient["default"].convertToType(data['links'], [_CreateAutomationLink["default"]]);
        }
      }

      return obj;
    }
  }]);

  return CreateAutomationRun;
}();
/**
 * Name of the new automation run.
 * @member {String} name
 */


CreateAutomationRun.prototype['name'] = undefined;
/**
 * Name of the source for the new automation run. If this source does not already exist in the target project, Testmo automatically creates a new one.  It is recommended to keep source names short. Good examples are `backend`, `frontend` or `mobile-iphone`.
 * @member {String} source
 */

CreateAutomationRun.prototype['source'] = undefined;
/**
 * Name of the configuration for the new automation run.
 * @member {String} config
 */

CreateAutomationRun.prototype['config'] = undefined;
/**
 * ID of the configuration for the new automation run. If both `config` and `config_id` are specified, `config_id` is given precedence.
 * @member {Number} config_id
 */

CreateAutomationRun.prototype['config_id'] = undefined;
/**
 * Name of the milestone for the new automation run.
 * @member {String} milestone
 */

CreateAutomationRun.prototype['milestone'] = undefined;
/**
 * ID of the milestone for the new automation run. If both `milestone` and `milestone_id` are specified, `milestone_id` is given precedence.
 * @member {Number} milestone_id
 */

CreateAutomationRun.prototype['milestone_id'] = undefined;
/**
 * List of tags for the new automation run. If a milestone in the same project has one or more matching automation tags, the new automation run is automatically linked to that milestone (unless `milestone` or `milestone_id` is also specified).
 * @member {Array.<String>} tags
 */

CreateAutomationRun.prototype['tags'] = undefined;
/**
 * List of externally stored test artifacts to link to the new automation run (such as log files, screenshots or test data).
 * @member {Array.<module:model/CreateAutomationArtifact>} artifacts
 */

CreateAutomationRun.prototype['artifacts'] = undefined;
/**
 * List of fields to attach to the new automation run (such as environment variables, errors or terminal output).
 * @member {Array.<module:model/CreateAutomationField>} fields
 */

CreateAutomationRun.prototype['fields'] = undefined;
/**
 * List of links to attach to the new automation run (such as a link back to the build in the CI tool that triggered the tests).
 * @member {Array.<module:model/CreateAutomationLink>} links
 */

CreateAutomationRun.prototype['links'] = undefined;
var _default = CreateAutomationRun;
exports["default"] = _default;
