// @testmo/testmo-api/dist/model/CreateAutomationArtifact.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _ApiClient = _interopRequireDefault(require("../ApiClient"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

/**
 * The CreateAutomationArtifact model module.
 * @module model/CreateAutomationArtifact
 * @version 1.0.0
 */
var CreateAutomationArtifact = /*#__PURE__*/function () {
  /**
   * Constructs a new <code>CreateAutomationArtifact</code>.
   * @alias module:model/CreateAutomationArtifact
   * @param name {String} Name or file name of the test artifact.
   * @param url {String} Link to an external resource to download the test artifact.
   */
  function CreateAutomationArtifact(name, url) {
    _classCallCheck(this, CreateAutomationArtifact);

    CreateAutomationArtifact.initialize(this, name, url);
  }
  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */


  _createClass(CreateAutomationArtifact, null, [{
    key: "initialize",
    value: function initialize(obj, name, url) {
      obj['name'] = name;
      obj['url'] = url;
    }
    /**
     * Constructs a <code>CreateAutomationArtifact</code> from a plain JavaScript object, optionally creating a new instance.
     * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
     * @param {Object} data The plain JavaScript object bearing properties of interest.
     * @param {module:model/CreateAutomationArtifact} obj Optional instance to populate.
     * @return {module:model/CreateAutomationArtifact} The populated <code>CreateAutomationArtifact</code> instance.
     */

  }, {
    key: "constructFromObject",
    value: function constructFromObject(data, obj) {
      if (data) {
        obj = obj || new CreateAutomationArtifact();

        if (data.hasOwnProperty('name')) {
          obj['name'] = _ApiClient["default"].convertToType(data['name'], 'String');
        }

        if (data.hasOwnProperty('note')) {
          obj['note'] = _ApiClient["default"].convertToType(data['note'], 'String');
        }

        if (data.hasOwnProperty('url')) {
          obj['url'] = _ApiClient["default"].convertToType(data['url'], 'String');
        }

        if (data.hasOwnProperty('mime_type')) {
          obj['mime_type'] = _ApiClient["default"].convertToType(data['mime_type'], 'String');
        }

        if (data.hasOwnProperty('size')) {
          obj['size'] = _ApiClient["default"].convertToType(data['size'], 'Number');
        }
      }

      return obj;
    }
  }]);

  return CreateAutomationArtifact;
}();
/**
 * Name or file name of the test artifact.
 * @member {String} name
 */


CreateAutomationArtifact.prototype['name'] = undefined;
/**
 * Short note or summary of the test artifact with a maximum of 80 characters.
 * @member {String} note
 */

CreateAutomationArtifact.prototype['note'] = undefined;
/**
 * Link to an external resource to download the test artifact.
 * @member {String} url
 */

CreateAutomationArtifact.prototype['url'] = undefined;
/**
 * MIME type of the test artifact (for example, `image/png`, `text/plain` or `application/octet-stream`).
 * @member {String} mime_type
 */

CreateAutomationArtifact.prototype['mime_type'] = undefined;
/**
 * File size of the test artifact in bytes.
 * @member {Number} size
 */

CreateAutomationArtifact.prototype['size'] = undefined;
var _default = CreateAutomationArtifact;
exports["default"] = _default;
