// @testmo/testmo-cli/dist/core/command.js
"use strict";
/*
 * Copyright (c) Testmo GmbH (Berlin, Germany)
 * All rights reserved.
 * contact@testmo.com - www.testmo.com
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Command = void 0;
var Command = /** @class */ (function () {
    function Command() {
        this.help = '';
        this.examples = '';
    }
    Command.prototype.trimOptionalArgument = function (value) {
        if (value === undefined) {
            return undefined;
        }
        else {
            value = this.trimRequiredArgument(value);
            if (value.length == 0) {
                return undefined;
            }
            else {
                return value;
            }
        }
    };
    Command.prototype.trimRequiredArgument = function (value) {
        return value.trim();
    };
    Command.prototype.trimOptionalArgumentList = function (value) {
        if (value === undefined) {
            return undefined;
        }
        else {
            return this.trimRequiredArgumentList(value);
        }
    };
    Command.prototype.trimRequiredArgumentList = function (value) {
        return value.map(function (v) {
            return v.trim();
        })
            .filter(function (v) {
            return v.length > 0;
        });
    };
    return Command;
}());
exports.Command = Command;
