// @testmo/testmo-cli/dist/controller/automationResourcesAddField.js
"use strict";
/*
 * Copyright (c) Testmo GmbH (Berlin, Germany)
 * All rights reserved.
 * contact@testmo.com - www.testmo.com
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutomationResourcesAddField = void 0;
var controller_1 = require("../core/controller");
var errors_1 = require("../lib/errors");
var resources_1 = require("../core/resources");
var io_1 = require("../lib/io");
var fs_1 = __importDefault(require("fs"));
var validator_1 = __importDefault(require("validator"));
var AutomationResourcesAddField = /** @class */ (function (_super) {
    __extends(AutomationResourcesAddField, _super);
    function AutomationResourcesAddField() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AutomationResourcesAddField.prototype.validate = function (options) {
        if (validator_1.default.isEmpty(options.name)) {
            throw new errors_1.ValidationError('Name cannot be empty');
        }
        if (!validator_1.default.isIn(options.type, ['string', 'url', 'text', 'console', 'html'])) {
            throw new errors_1.ValidationError('Type must be either: string, url, text, console, html');
        }
        if (options.type == 'url' && options.value !== undefined &&
            !validator_1.default.isEmpty(options.value)) {
            if (!validator_1.default.isURL(options.value, {
                require_tld: false,
                require_protocol: true
            })) {
                throw new errors_1.ValidationError('URL is not a valid link');
            }
        }
    };
    AutomationResourcesAddField.prototype.run = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var value, stdin, resources;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.logControllerOptions(options);
                        this.validate(options);
                        value = options.value;
                        if (!(options.type == 'text' ||
                            options.type == 'console' ||
                            options.type == 'html')) return [3 /*break*/, 2];
                        return [4 /*yield*/, (0, io_1.readFromStdin)()];
                    case 1:
                        stdin = _a.sent();
                        if (stdin === false || typeof stdin !== 'string') {
                            throw new Error('Required stdin for text, console, html is empty');
                        }
                        value = stdin.trim();
                        return [3 /*break*/, 3];
                    case 2:
                        if (value === undefined) {
                            throw new errors_1.ValidationError('Field value is missing');
                        }
                        _a.label = 3;
                    case 3:
                        resources = new resources_1.Resources();
                        if (fs_1.default.existsSync(options.resourceFile)) {
                            resources.readFromFile(options.resourceFile);
                        }
                        resources.addField({
                            type: options.type,
                            name: options.name,
                            value: value
                        });
                        resources.writeToFile(options.resourceFile);
                        return [2 /*return*/];
                }
            });
        });
    };
    return AutomationResourcesAddField;
}(controller_1.Controller));
exports.AutomationResourcesAddField = AutomationResourcesAddField;
