// @testmo/testmo-api/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ApiClient", {
  enumerable: true,
  get: function get() {
    return _ApiClient["default"];
  }
});
Object.defineProperty(exports, "AppendToAutomationRun", {
  enumerable: true,
  get: function get() {
    return _AppendToAutomationRun["default"];
  }
});
Object.defineProperty(exports, "AppendToAutomationRunThread", {
  enumerable: true,
  get: function get() {
    return _AppendToAutomationRunThread["default"];
  }
});
Object.defineProperty(exports, "AutomationRunsApi", {
  enumerable: true,
  get: function get() {
    return _AutomationRunsApi["default"];
  }
});
Object.defineProperty(exports, "CompleteAutomationRun", {
  enumerable: true,
  get: function get() {
    return _CompleteAutomationRun["default"];
  }
});
Object.defineProperty(exports, "CompleteAutomationRunThread", {
  enumerable: true,
  get: function get() {
    return _CompleteAutomationRunThread["default"];
  }
});
Object.defineProperty(exports, "CreateAutomationArtifact", {
  enumerable: true,
  get: function get() {
    return _CreateAutomationArtifact["default"];
  }
});
Object.defineProperty(exports, "CreateAutomationField", {
  enumerable: true,
  get: function get() {
    return _CreateAutomationField["default"];
  }
});
Object.defineProperty(exports, "CreateAutomationLink", {
  enumerable: true,
  get: function get() {
    return _CreateAutomationLink["default"];
  }
});
Object.defineProperty(exports, "CreateAutomationRun", {
  enumerable: true,
  get: function get() {
    return _CreateAutomationRun["default"];
  }
});
Object.defineProperty(exports, "CreateAutomationRunThread", {
  enumerable: true,
  get: function get() {
    return _CreateAutomationRunThread["default"];
  }
});
Object.defineProperty(exports, "CreateAutomationTest", {
  enumerable: true,
  get: function get() {
    return _CreateAutomationTest["default"];
  }
});
Object.defineProperty(exports, "CreatedAutomationRun", {
  enumerable: true,
  get: function get() {
    return _CreatedAutomationRun["default"];
  }
});
Object.defineProperty(exports, "CreatedAutomationRunThread", {
  enumerable: true,
  get: function get() {
    return _CreatedAutomationRunThread["default"];
  }
});
Object.defineProperty(exports, "ErrorCollection", {
  enumerable: true,
  get: function get() {
    return _ErrorCollection["default"];
  }
});
Object.defineProperty(exports, "User", {
  enumerable: true,
  get: function get() {
    return _User["default"];
  }
});
Object.defineProperty(exports, "UserApi", {
  enumerable: true,
  get: function get() {
    return _UserApi["default"];
  }
});

var _ApiClient = _interopRequireDefault(require("./ApiClient"));

var _AppendToAutomationRun = _interopRequireDefault(require("./model/AppendToAutomationRun"));

var _AppendToAutomationRunThread = _interopRequireDefault(require("./model/AppendToAutomationRunThread"));

var _CompleteAutomationRun = _interopRequireDefault(require("./model/CompleteAutomationRun"));

var _CompleteAutomationRunThread = _interopRequireDefault(require("./model/CompleteAutomationRunThread"));

var _CreateAutomationArtifact = _interopRequireDefault(require("./model/CreateAutomationArtifact"));

var _CreateAutomationField = _interopRequireDefault(require("./model/CreateAutomationField"));

var _CreateAutomationLink = _interopRequireDefault(require("./model/CreateAutomationLink"));

var _CreateAutomationRun = _interopRequireDefault(require("./model/CreateAutomationRun"));

var _CreateAutomationRunThread = _interopRequireDefault(require("./model/CreateAutomationRunThread"));

var _CreateAutomationTest = _interopRequireDefault(require("./model/CreateAutomationTest"));

var _CreatedAutomationRun = _interopRequireDefault(require("./model/CreatedAutomationRun"));

var _CreatedAutomationRunThread = _interopRequireDefault(require("./model/CreatedAutomationRunThread"));

var _ErrorCollection = _interopRequireDefault(require("./model/ErrorCollection"));

var _User = _interopRequireDefault(require("./model/User"));

var _AutomationRunsApi = _interopRequireDefault(require("./api/AutomationRunsApi"));

var _UserApi = _interopRequireDefault(require("./api/UserApi"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
